syntax = "proto3";

package descarteslabs.workflows;

import "google/protobuf/timestamp.proto";
import "google/protobuf/wrappers.proto";

import "descarteslabs/common/proto/destinations/destinations.proto";
import "descarteslabs/common/proto/formats/formats.proto";
import "descarteslabs/common/proto/types/types.proto";
import "descarteslabs/common/proto/typespec/typespec.proto";
// backwards compatible fix for ErrorCode
import public "descarteslabs/common/proto/errors/errors.proto";

service JobAPI {
  rpc CreateJob(CreateJobRequest) returns (Job) {}
  rpc ListJobs(ListJobsRequest) returns (stream Job) {}
  rpc GetJob(GetJobRequest) returns (Job) {}
  rpc CancelJob(CancelJobRequest) returns (CancelJobResponse) {}
  rpc WatchJob(WatchJobRequest) returns (stream Job.State) {}
}

message Job {
  reserved 4;

  // Core contents: ID, graft, type, and arguments

  string id = 1;
  string serialized_graft = 2;
  Typespec typespec = 3;

  // The arguments to the graft function in `serialized_graft`.
  // Keys are parameter names. Values are argument values,
  // as either JSON-serialized literals or JSON-serialized grafts.
  map<string, string> arguments = 15;

  // Env variables

  string geoctx_graft = 16;
  // internal use only---ignored and always returned as null
  string token = 17;
  bool no_ruster = 18;

  // Internal execution settings & metadata

  string channel = 6;
  string client_version = 14;
  bool no_cache = 5;
  bool trace = 19;
  // milliseconds since the UNIX epoch (UTC)
  int64 timestamp = 7;
  string user = 8;
  string org = 9;

  google.protobuf.Timestamp expires_timestamp = 20;

  // State information

  enum Stage {
    UNSPECIFIED = 0;
    QUEUED = 1;
    PREPARING = 2;
    RUNNING = 3;
    SAVING = 4;
    FAILED = 5;
    SUCCEEDED = 6;
    CANCELLED = 7;
  }

  message TasksProgress {
    google.protobuf.UInt32Value waiting = 1;
    google.protobuf.UInt32Value ready = 2;
    google.protobuf.UInt32Value running = 3;
    google.protobuf.UInt32Value finished = 4;
  }

  message Error {
    ErrorCode code = 1;
    string message = 2;
  }

  message State {
    Stage stage = 1;
    TasksProgress tasks_progress = 2;
    Error error = 3;
    // milliseconds since the UNIX epoch (UTC)
    int64 timestamp = 4;
  }

  State state = 10;

  // Output settings

  ResultType type = 11;
  Format format = 12;
  Destination destination = 13;
}

message CreateJobRequest {
  reserved 3;

  // Core contents: graft, type, and arguments

  string serialized_graft = 1;
  Typespec typespec = 2;
  map<string, string> arguments = 10;

  // Env variables

  string geoctx_graft = 11;
  bool no_ruster = 12;

  // Internal execution settings

  string channel = 5;
  string client_version = 9;
  bool no_cache = 4;
  bool trace = 13;

  // Output settings

  ResultType type = 6;
  Format format = 7;
  Destination destination = 8;
}

message GetJobRequest { string id = 1; }

message CancelJobRequest { string id = 1; }

message CancelJobResponse {}

message WatchJobRequest { string id = 1; }

message ListJobsRequest {
  string start_datetime = 1;
  string end_datetime = 2;
}
